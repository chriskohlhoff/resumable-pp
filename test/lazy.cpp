#include <stdio.h>
#include <utility>

class heavy
{
public:
  heavy(const char* s) { printf("heavy(%s)\n", s); }
  ~heavy() { printf("~heavy()\n"); }
  void print() { printf("here we are\n"); }
};

int main()
{
  const char* text = "hello";

  auto lazy_h = [=]() resumable -> heavy* {
    for (heavy h(text);;)
      yield &h;
  };

  printf("before use\n");
  lazy_h()->print();
  printf("after use\n");
}
