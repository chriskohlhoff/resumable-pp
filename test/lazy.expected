before use
heavy(hello)
here we are
after use
~heavy()
